mountpoint-s3-crt/src/io/stream.rs

```diff
@@ -0,0 +1,192 @@
+        // cause it to live longer than expected. (Or just kill this function...)
+        let inner = unsafe { aws_input_stream_new_from_cursor(allocator.inner.as_ptr(), &cursor).ok_or_last_error()? };
+
+        Ok(Self {
+            inner,
+            _phantom: Default::default(),
+        })
+    }
+}
+
+impl<'a> Drop for InputStream<'a> {
+    fn drop(&mut self) {
+        // SAFETY: self.inner is a valid `aws_input_stream`.
+        unsafe {
+            aws_input_stream_release(self.inner.as_ptr());
+        }
+    }
+}
+
+impl<'a> InputStream<'a> {
+    /// Seek to the given offset. Basis is either BEGIN or END, and describes where to seek from.
+    pub fn seek(&self, offset: i64, basis: SeekBasis) -> Result<(), Error> {
+        // SAFETY: self.inner is a valid input stream.
+        unsafe { aws_input_stream_seek(self.inner.as_ptr(), offset, basis.into()).ok_or_last_error() }
+    }
+
+    /// Read some data into `buffer`, and return how many bytes were read.
+    pub fn read(&self, buffer: &mut [u8]) -> Result<usize, Error> {
+        let mut byte_buf = aws_byte_buf {
+            len: 0,
+            buffer: buffer.as_mut_ptr(),
+            capacity: buffer.len(),
+            allocator: std::ptr::null_mut(),
+        };
+
+        // SAFETY: we know that the aws_byte_buf we just made points to a valid buffer, and trust
+        // the CRT function not to write outside that buffer's capacity. Also, self.inner is a
+        // valid input stream.
+        unsafe {
+            aws_input_stream_read(self.inner.as_ptr(), &mut byte_buf).ok_or_last_error()?;
+        };
+
+        assert_eq!(byte_buf.capacity, buffer.len(), "capacity should not change");
+
+        assert!(
+            byte_buf.len <= buffer.len(),
+            "should not have written more than available"
+        );
+        Ok(byte_buf.len)
+    }
+
+    /// Get the status of this stream. Can be used to indicate if stream is at the EOF.
+    pub fn get_status(&self) -> Result<StreamStatus, Error> {
+        let mut status: aws_stream_status = Default::default();
+
+        // SAFETY: self.inner is a valid input stream and status is a local variable.
+        unsafe {
+            aws_input_stream_get_status(self.inner.as_ptr(), &mut status).ok_or_last_error()?;
+        }
+
+        Ok(status.into())
+    }
+
+    /// Get the length of this input stream, in bytes. If a length cannot be determined, return Err.
+    pub fn get_length(&self) -> Result<usize, Error> {
+        let mut out_length: i64 = 0;
+
+        // SAFETY: self.inner is a valid input stream and out_length is a pointer to a local variable.
+        unsafe {
+            aws_input_stream_get_length(self.inner.as_ptr(), &mut out_length).ok_or_last_error()?;
+        }
+
+        Ok(out_length.try_into().expect("failed to convert i64 to usize"))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::common::allocator::Allocator;
+
+    #[test]
+    fn test_slice_cursor() {
```

Comment:
Can we have another test that does partial read from the stream and verify it doesn't end prematurely?